### PR TITLE
feat: persist ExpectedUtxo to disk (db)

### LINKED ISSUE #172: ExpectedUtxo not persisted, can lead to loss of funds.
disclaimer:  I'm making a preliminary report so this doesn't get forgotten. I intend to make a test case to verify this issue is real, but from reading through relevant code I believe it is.

disclaimer2: this may be a known todo item on a list somewhere. 

If I've missed something, please correct and close if appropriate.

summary: if a node is stopped after sending a transaction (using off-chain utxo notification for change) but before the transaction is confirmed in a block then the utxo sent to change address is lost and can never be claimed.

steps to reproduce:   (I have not done this yet)

 1. check initiial wallet balance and record it.
 2. send funds to 3rd party, with some change going back to own wallet.
 3. shutdown wallet.
 4. wait for a block to be mined by another node.
 5. restart node.
 6. ensure tip of node has the mined block.
 7.  check wallet balance.

correct outcome: the balance should be original balance - send amount - fee.
predicted outcome:  the balance will be original balance - send amount - change amount - fee.

why would this happen?

because ExpectedUtxo are not persisted to disk, only MonitoredUtxo.   The former are converted into the latter after a block is mined.   So ExpectedUtxo live only in RAM (inside UtxoNotificationPool) and there exists a window of time during which the ExpectedUtxo can be lost if the node is shutdown.  This could be as bad as losing 100% of wallet funds.

Example scenario:

1. Alice starts her node, syncs, sends funds Bob, and immediately shuts down.
2. A week later, Alice starts her node again and is surprised that her wallet is now lower than expected.

This seems likely to be a common usage pattern.  (and even more so for light nodes)

Suggested fix:   We might consider adding a field expected_utxos: DbtVec<ExpectedUtxos> into RustyWalletDatabase.  This would store expected_utxos alongside monitored_utxos in the wallet db.   Alternatively a separate DB file could be utilized within UtxoNotificationPool.

### PR DESCRIPTION
**Pull request by dan-da**
closes #172, #137.

fixes loss-of-funds scenario.

The primary change here is to move storage of `expected_utxos` from `UtxoNotificationPool` into RustyWalletDatabase.  Thus `expected_utxos` now exists as a DbtVec in the same wallet database as monitored_utxos.  `UtxoNotificationPool` is removed entirely.  Also pruning of stale expected utxos has been temporarily disabled.

Commentary:
 1. storing expected_utxos in the same DB as monitored_utxos causes DB writes to be atomic across both data sets which is desirable.  Also, there is now a nice symmetry for both these utxo data sets.
 2. An alternate design would be to use a separate levelDB file for expected_utxos.  This would enable deletion of individual entries, but we would lose transactional guarantee with other items in the wallet db.
 3. UtxoNotificationPool had code related to PeerNotification that was not actually being used.  I believe this may have been implemented to support planned future functionality, but I think it best to remove it now, until such time as something like that is needed.
 4. In my judgement, it is dangerous to ever expire expected utxos as this could result in a loss of funds, for example during a deep re-org.  I think a thorough analysis should be performed to determine rules for when it is safe, if ever, to expire/prune expected utxos.  So until that is done, I have disabled the call to prune, although a method still exists.  See my code comments for WalletState::prune_stale_expected_utxos().  Pruning is essentially an optimization, so I think it is safest to launch without it and optimize if/when the time is right.
 5. The prune method itself had to be modified to first clear all entries, then add back non-stale entries.  This is because DbtVec does not support removing individual entries, except for the last.
 6. In my view, the code is simpler and cleaner now with UtxoNotificationPool gone.  If this change is controversial, let me know.

changes:
 * add RustyWalletDatabase::expected_utxos (DbtVec)
 * remove UtxoNotificationPool
 * remove UtxoNotifier::PeerUnsigned
 * add WalletState methods: add_expected_utxo(), scan_for_expected_utxos()
 * disables timer call to prune_stale_expected_utxos() for now.
 + remove cli args: max_utxo_notification_size, max_unconfirmed_utxo_notification_count_per_peer
 * mod utxo_notification_pool --> expected_utxo
 * derive Hash for Timestamp

tests:
 * adapt existing tests to changes
 * move tests from utxo_notification_pool into wallet_state
 * adds regression tests for issue #172 that verify:
    1.  expected_utxo are persisted if db is written to disk.
    2.  expected_utxo are not persisted if db is not written to disk.

### CODE REVIEW

**Review by aszepieniec — APPROVED** [review]
LGTM. Minor, mostly superficial suggestions.

File: src/models/state/wallet/expected_utxo.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    prelude::twenty_first,
+    util_types::mutator_set::{addition_record::AdditionRecord, commit},
+};
+use get_size::GetSize;
+use serde::{Deserialize, Serialize};
+use twenty_first::{math::tip5::Digest, util_types::algebraic_hasher::AlgebraicHasher};
+
+// 28 days in secs
+pub const UNRECEIVED_UTXO_NOTIFICATION_THRESHOLD_AGE_IN_SECS: u64 = 28 * 24 * 60 * 60;
+pub const RECEIVED_UTXO_NOTIFICATION_THRESHOLD_AGE_IN_SECS: u64 = 3 * 24 * 60 * 60;
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs:13]: thx. these were only used inside WalletState::prune_stale_expected_utxos(), so I am just moving them in there.

File: src/models/state/wallet/expected_utxo.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+/// represents utxo and secrets necessary for recipient to claim it.
+///
+/// [ExpectedUtxo] is intended for offchain temporary storage of utxos that a
+/// wallet sends to itself, eg change outputs.
+///
+/// The `ExpectedUtxo` will exist in the local [UtxoNotificationPool] from the
+/// time the transaction is sent until it is mined in a block and claimed by the
+/// wallet.
+///
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs:23]: fixed.

File: src/models/state/wallet/expected_utxo.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+/// An `ExpectedUtxo` really only needs `utxo`, `sender_randomness` and
+/// `receiver_identifier`.  These match the fields used in `PublicAnnouncement`.
+///
+/// However it then becomes necessary to scan all known wallet keys (of all key
+/// types) in order to find the matching key, in order to obtain the preimage.
+///
+/// improvement: To avoid scanning a map of \[`receiver_identifier`\] -->
+/// `derivation_index` could be stored in local wallet state for all known keys.
+///
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs:47]: moved into #176 with a link from the doc-comment.

File: src/models/state/wallet/expected_utxo.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+/// However no such map presently exists, and so the most efficient and easy
+/// thing is to include the preimage in [ExpectedUtxo] instead of a
+/// `receiver_identifier`.  This buys us an equivalent optimization for little
+/// effort.
+///
+/// #### compromise
+///
+/// Because the preimage is necessary to create an [ExpectedUtxo]
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs:55]: moved into #176 with a link from the doc-comment.

File: src/models/state/wallet/expected_utxo.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+///
+/// Further if [ExpectedUtxo] and `PublicAnnouncement` use the same fields then
+/// they can share much of the same codepath when claiming.  At present, we have
+/// separate codepaths that perform largely the same function.
+///
+/// We may revisit this area in the future, as it seems ripe for improvement.
+/// In particular wallet receiver_identifier map idea indicates it is possible
+/// to do this efficiently.  As such it may be best to implement at least the
+/// scanning based approach before mainnet.
+///
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs:68]: moved into #176 with a link from the doc-comment.

File: src/models/state/wallet/expected_utxo.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+/// A branch with an implementation of the scanning approach exists:
+/// danda/symmetric_keys_and_expected_utxos_without_receiver_preimage
+///
+/// At time of writing many tests are not passing and need updating with the new
+/// field.
+///
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs:74]: I think we could still simplify/unify the code a bit more with the changes in this branch yes, though not a top priority.  

anyway, its moved into #176 with a link from the doc-comment.

File: src/models/state/wallet/expected_utxo.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+/// thing is to include the preimage in [ExpectedUtxo] instead of a
+/// `receiver_identifier`.  This buys us an equivalent optimization for little
+/// effort.
+///
+/// #### compromise
+///
+/// Because the preimage is necessary to create an [ExpectedUtxo]
+/// `create_transaction()` must accept a `change_key: SpendingKey` parameter
+/// rather than `change_address: ReceivingAddress`.  One would normally expect
+/// an output to require only an address.
```
[the remark this answers is not in the document]
Comment (dan-da) [reply src/models/state/wallet/expected_utxo.rs]: added a note to the comment which is now in #176.

**Comment by dan-da** [comment]
ok, I addressed all the suggestions in 3c049fb

**Comment by Sword-Smith** [comment]
It seems like a simplification to use the database for storing expected UTXOs instead of using a mempool-like data structure that I originally implemented this as. With the mempool-like structure you'd have to have two different sets of logic for persisting and handling the UTXO notification pool in RAM.

Some functionality that's lost, though, is that the `UtxoNotifier` can no longer be a peer. And, when the `UtxoNotifier` can be a peer, it makes sense to have a "credibility" value attached to `ExpectedUtxo`. And then it makes sense to have some concept of ordering of the UTXO notifications. Or, at least to have some kind of DOS protection to prevent a peer from using all your available disk and RAM for UTXO notifications.

The transaction notifications that you can put on-chain through public announcements both cost transaction fees and they leak information. If we are to have a way of transacting that avoids these two downsides, I'd say it makes sense to allow the protocol to share UTXO notifications somehow, a notification being the `(Utxo, Digest)` tuple, representing the UTXO and the sender randomness. We'd certainly want to encrypt this information if 3rd parties see it.

You don't have two answer this latter question now, about how to share UTXO notifications through `neptune-core` without those notifications showing up in the block. But the `ExpectedUtxo` data structure/the wallet state should somehow support this use case.

**Review by Sword-Smith — CHANGES_REQUESTED** [review]
It seems like a simplification to use the database for storing expected UTXOs instead of using a mempool-like data structure that I originally implemented this as. With the mempool-like structure you'd have to have two different sets of logic for persisting and handling the UTXO notification pool in RAM.

Some functionality that's lost, though, is that the UtxoNotifier can no longer be a peer. And, when the UtxoNotifier can be a peer, it makes sense to have a "credibility" value attached to ExpectedUtxo. And then it makes sense to have some concept of ordering of the UTXO notifications. Or, at least to have some kind of DOS protection to prevent a peer from using all your available disk and RAM for UTXO notifications.

The transaction notifications that you can put on-chain through public announcements both cost transaction fees and they leak information. If we are to have a way of transacting that avoids these two downsides, I'd say it makes sense to allow the protocol to share UTXO notifications somehow, a notification being the (Utxo, Digest) tuple, representing the UTXO and the sender randomness. We'd certainly want to encrypt this information if 3rd parties see it.

You don't have two answer this latter question now, about how to share UTXO notifications through neptune-core without those notifications showing up in the block. But the ExpectedUtxo data structure/the wallet state should somehow support this use case.

**Comment by aszepieniec** [comment]
@dan-da You mentioned during the standup call on Monday that your next task would be to implement support for off-band UTXO notifications. Could you elaborate on how you expect that to work and how it is supposed to integrate with the changes introduced by this PR (assuming it is merged)? Maybe you and @Sword-Smith aren't that far apart when it comes to concrete next steps.

I'm fine with a CLI-only interface for off-band UTXO notifications as an immediate target. We probably do want to support an automatic peer-to-peer overlay protocol at some point in the future, but that needs to come after an elaborate design process with DOS-resistance built in. That's where I stand now, but that might change based on arguments.

**Comment by dan-da** [comment]
> You mentioned during the standup call on Monday that your next task would be to implement support for off-band UTXO notifications.  Could you elaborate on how you expect that to work and how it is supposed to integrate with the changes introduced by this PR (assuming it is merged)? Maybe you and @Sword-Smith aren't that far apart when it comes to concrete next steps.

Sure.  I view off-chain notifications as being orthogonal / complementary to this PR or to a future p2p shared utxo pool.

|name|send to|on/off chain|middle-man|
|-------|----------|------|---------------|
|this PR|self|off-chain|neptune-core|
|p2p pool|others|off-chain|neptune-core|
|out-of-band|others|off-chain|email, sms, signal, etc|


Basically the out-of-band process would look something like:

1. add UtxoNotification::OffChainSerialized(UtxoTransfer). Where UtxoTransfer is a placeholder name for a serializable struct with same fields that are in the PublicAnnouncement BFieldElements.

2. modify send_to_many() or add new RPCs such that caller can specify TxOutputList directly, which includes UtxoNotification for each output.  Caller can use a helper method on GlobalState to generate these from [(ReceivingAddress, amount)] and map from ReceivingAddress to each UtxoTransfer.

3. sender then provides the serialized UtxoTransfer(s) to recipient(s) off-chain and out-of-band.  (not neptune-core's concern how this occurs)

4. recipient imports/claims utxos via a new claim_utxos() RPC.  The UtxoTransfer is stored permanently in recipient's local state in case blockchain re-scan is ever needed. 

 5. ideally there would be a way (RPC call) for sender to regenerate the data in case it is lost in transit somehow.  I haven't really given that any thought yet.

**Comment by Sword-Smith** [comment]
For this PR, I see a danger of some unfortunate scaling behavior, specifically in the `WalletState::scan_for_expected_utxos`

```rust
    pub async fn scan_for_expected_utxos<'a>(
        &'a self,
        transaction: &'a Transaction,
    ) -> impl Iterator<Item = AnnouncedUtxo> + 'a {
        let expected_utxos = self.wallet_db.expected_utxos().get_all().await;

        transaction.kernel.outputs.iter().filter_map(move |a| {
            expected_utxos
                .iter()
                .find(|eu| eu.addition_record == *a)
                .map(|eu| eu.into())
        })
    }
```

Here, the `expected_utxos` variable grows indefinitely with the number of blocks that have been mined locally. And the complexity of this function scales as $N \cdot M$, where $N$ is the number of expected UTXOs in the wallet state and $M$ the number of outputs in the block. Having a method that only fetches the not-yet-confirmed expected UTXOs would fix this issue.

**Comment by dan-da** [comment]
> Here, the `expected_utxos` variable grows indefinitely with the number of blocks that have been mined locally. 

good catch.

yes this is yet another limitation of using DbtVec.  We lose the ability to lookup by the AdditionRecord hash, which is what the code in `master` does.

> Having a method that only fetches the not-yet-confirmed expected UTXOs would fix this issue.

In this PR, all ExpectedUtxo are stored in a single DbtVec, as opposed to having separate DbtVec for confirmed vs not-confirmed.  I see a few possible ways to be more efficient.

1. generate an index on the fly.  At the start of scan_for_expected_utxos()  loop through all ExpectedUtxo and add them to a HashMap indexed by AdditionRecord.  This is o(n) with number of expected_utxos to generate the index plus o(m) with number of tx outputs.

2. add a second DbtVec to the database that contains indexes of the not-yet-confirmed ExpectedUtxo.  Then only fetch ExpectedUtxo at these indices.  This is o(n*m) with the number of not-confirmed utxos * number of tx outputs.

3. use a separate leveldb file dedicated to ExpectedUtxo.  Then we simply store the ExpectedUtxo with the AdditionRecord hash as the key.  This is o(m) with the number of tx outputs.

4. use a database such as `redb` that supports table-like namespaces so that we can use AdditionRecord hash as the key but still put them in the wallet DB.  This is also o(m) with number of tx outputs.

long term, my preference is (4).

With leveldb:
  (1) is easy/fast to impl, but still grows o(n)+o(m).   probably "good enough" for mainnet.   unless/until wallets have tens of thousands of expected_utxos.
  (2) is a bit gross as it requires a 2nd dbtvec in the DB that must be maintained.  It is still o(n*m), but the size of n is much reduced.
  (3) is the most efficient at o(m), but we lose atomicity with the rest of the wallet db.  I haven't thought deeply if this is any problem or not.  But if nothing else, it is a possible complication when restoring wallet backups.

I would suggest (1) for now, and consider (4) later.

**Comment by dan-da** [comment]
ok, I impl'd (1) in 3776c78a66903f7a0123716904bf7082cb9bfda1.    @Sword-Smith what do you think?

**Comment by Sword-Smith** [comment]
> ok, I impl'd (1) in [3776c78](https://github.com/Neptune-Crypto/neptune-core/commit/3776c78a66903f7a0123716904bf7082cb9bfda1). @Sword-Smith what do you think?

Looks good. This illustrates, though, that it's important that we log the time to append a new block to the state, such that we can catch such problems in the future.

**Review by Sword-Smith — APPROVED** [review]

**Pull request merged**